File: python/othello/player.py
"""
Class Player
Defines one player for Othello
Akseli Lukkarila
2019-2026
"""

from __future__ import annotations

import random
import time

from enum import IntEnum
from typing import Self

try:
    from othello.board import Board
    from othello.colorprint import Color, print_color, print_error
    from othello.models import Disk, Move, Square
    from othello.settings import PlayerSettings
except ModuleNotFoundError:
    from board import Board
    from colorprint import Color, print_color, print_error
    from models import Disk, Move, Square
    from settings import PlayerSettings


class Player:
    """Defines one player that can be either human or computer controlled."""

    def __init__(self, disk: Disk, settings: PlayerSettings | None = None):
        """Initialize new player for the given disk color."""
        self.can_play: bool = True
        self._disk: Disk = disk
        self._player_type: PlayerType = PlayerType.HUMAN
        self._rounds_played: int = 0
        self._settings: PlayerSettings = (
            settings if settings is not None else PlayerSettings.default()
        )

    @classmethod
    def black(cls, settings: PlayerSettings) -> Self:
        """Shorthand to initialize a new player for black disks."""
        return cls(Disk.BLACK, settings)

    @classmethod
    def white(cls, settings: PlayerSettings) -> Self:
        """Shorthand to initialize a new player for white disks."""
        return cls(Disk.WHITE, settings)

    def play_one_move(self, board: Board) -> str | None:
        """Play one round as this player."""
        if not self._settings.check_mode:
            print(f"Turn: {self._disk}")

        moves = board.possible_moves(self._disk)
        if not moves:
            self.can_play = False
            if not self._settings.check_mode:
                print_color("  No moves available...", Color.yellow)

            return None

        self.can_play = True
        if self.human() and self._settings.show_helpers and not self._settings.check_mode:
            board.print_possible_moves(moves)

        chosen_move = (
            self._get_human_move(moves) if self.human() else self._get_computer_move(moves)
        )
        board.place_disk(chosen_move)
        if not self._settings.check_mode:
            board.print_score()
        self._rounds_played += 1
        if not self._settings.test_mode:
            time.sleep(1)

        return chosen_move.log_entry()

    def reset(self) -> None:
        """Reset player status for a new game."""
        self.can_play = True
        self._rounds_played = 0

    def human(self) -> bool:
        """Returns true if player is controlled by a human player."""
        return self._player_type.human()

    def computer(self) -> bool:
        """Returns true if player is controlled by computer."""
        return self._player_type.computer()

    def set_player_type(self, player_type: PlayerType) -> None:
        """Set the player as human or computer controlled."""
        self._player_type = player_type

    def set_human(self) -> None:
        """Set the player as human controlled."""
        self.set_player_type(PlayerType.HUMAN)

    def set_computer(self) -> None:
        """Set the player as computer controlled."""
        self.set_player_type(PlayerType.COMPUTER)

    def _get_computer_move(self, moves: list[Move]) -> Move:
        """Return move chosen by computer."""
        if not self._settings.check_mode:
            print("  Computer plays...")

        if self._settings.test_mode:
            chosen_move = moves[0]
        else:
            # Wait a bit and pick a random move
            time.sleep(random.uniform(1.0, 2.0))
            chosen_move = random.choice(moves)

        if not self._settings.check_mode:
            print(f"  {chosen_move.square} -> {chosen_move.value}")

        return chosen_move

    def _get_human_move(self, moves: list[Move]) -> Move:
        """Return move chosen by a human player."""
        while True:
            square = self._get_square()
            # Check that the chosen square is actually one of the possible moves
            move = next((move for move in moves if move.square == square), None)
            if move:
                return move

            print_error(f"  Can't place a {self._disk} disk in square {square}!")

    @staticmethod
    def _get_square() -> Square:
        """Ask human player for square coordinates."""
        while True:
            pos = input("  Give disk position (x,y): ")
            values = pos.strip().split(",")
            if len(values) == 2:
                try:
                    x = int(values[0])
                    y = int(values[1])
                except ValueError:
                    x, y = -1, -1
                if x >= 0 and y >= 0:
                    return Square(x, y)

            print_error("  Give coordinates in the form 'x,y'!")

    def type_string(self) -> str:
        """Return player type description string."""
        return str(self._player_type)

    def __str__(self) -> str:
        return f"{self._disk} | {self.type_string()} | Moves: {self._rounds_played}"


class PlayerType(IntEnum):
    """Player can be controlled either by a human or computer."""

    HUMAN = 0
    COMPUTER = 1

    def human(self) -> bool:
        """Check if the player is controlled by a human."""
        return self == PlayerType.HUMAN

    def computer(self) -> bool:
        """Check if the player is controlled by a computer."""
        return self == PlayerType.COMPUTER

    def __str__(self) -> str:
        return "Human   " if self == PlayerType.HUMAN else "Computer"

File: csharp/Othello/Player.cs
//==========================================================
// Class Player
// Defines one player for Othello
// Akseli Lukkarila
// 2019-2026
//==========================================================

using System;
using System.Collections.Generic;
using System.Drawing;
using System.Globalization;
using System.Threading;

namespace Othello {
    /// <summary>Defines one player that can be either human or computer controlled.</summary>
    internal sealed class Player(Disk disk, PlayerSettings settings) {
        public bool CanPlay = true;
        internal Disk Disk { get; } = disk;
        private PlayerType _playerType = PlayerType.Human;
        internal int RoundsPlayed;
        internal readonly PlayerSettings Settings = settings;
        private readonly Random _random = new();

        /// <summary>Shorthand to initialize a new player for black disks.</summary>
        public static Player Black(PlayerSettings settings) {
            return new Player(Disk.Black, settings);
        }

        /// <summary>Shorthand to initialize a new player for white disks.</summary>
        public static Player White(PlayerSettings settings) {
            return new Player(Disk.White, settings);
        }

#nullable enable
        /// <summary>Play one round as this player.</summary>
        public string? PlayOneMove(Board board) {
            if (!Settings.CheckMode) {
                Console.WriteLine($"Turn: {Disk.DiskString()}");
            }
            List<Move> moves = board.PossibleMoves(Disk);
            if (moves.Count == 0) {
                CanPlay = false;
                if (!Settings.CheckMode) {
                    ColorPrint.WriteLine("  No moves available...", Color.Yellow);
                }
                return null;
            }

            CanPlay = true;
            if (Human() && Settings is { ShowHelpers: true, CheckMode: false }) {
                board.PrintPossibleMoves(moves);
            }
            Move chosenMove = Human() ? GetHumanMove(moves) : GetComputerMove(moves);
            board.PlaceDisk(chosenMove);
            if (!Settings.CheckMode) {
                board.PrintScore();
            }
            ++RoundsPlayed;
            if (!Settings.TestMode) {
                Thread.Sleep(1000);
            }
            return chosenMove.LogEntry();
        }
#nullable disable

        /// <summary>Reset player status for a new game.</summary>
        public void Reset() {
            CanPlay = true;
            RoundsPlayed = 0;
        }

        /// <summary>Returns true if player is controlled by a human player.</summary>
        public bool Human() {
            return _playerType.Human();
        }

        /// <summary>Returns true if player is controlled by computer.</summary>
        public bool Computer() {
            return _playerType.Computer();
        }

        /// <summary>Set the player as human or computer controlled.</summary>
        public void SetPlayerType(PlayerType playerType) {
            _playerType = playerType;
        }

        /// <summary>Set the player as human controlled.</summary>
        public void SetHuman() {
            SetPlayerType(PlayerType.Human);
        }

        /// <summary>Set the player as computer controlled.</summary>
        public void SetComputer() {
            SetPlayerType(PlayerType.Computer);
        }

        /// <summary>Return move chosen by computer.</summary>
        private Move GetComputerMove(List<Move> moves) {
            if (!Settings.CheckMode) {
                Console.WriteLine("  Computer plays...");
            }
            Move chosenMove;
            if (Settings.TestMode) {
                chosenMove = moves[0];
            } else {
                // Wait a bit and pick a random move
                Thread.Sleep(_random.Next(1000, 2000));
                chosenMove = moves[_random.Next(moves.Count)];
            }
            if (!Settings.CheckMode) {
                Console.WriteLine($"  {chosenMove.Square} -> {chosenMove.Value}");
            }
            return chosenMove;
        }

        /// <summary>Return move chosen by a human player.</summary>
        private Move GetHumanMove(List<Move> moves) {
            while (true) {
                Square square = GetSquare();
                // Check that the chosen square is actually one of the possible moves
                if (moves.Exists(move => square.Equals(move.Square))) {
                    return moves.Find(move => square.Equals(move.Square));
                }
                ColorPrint.PrintError($"  Can't place a {Disk.DiskString()} disk in square {square}!");
            }
        }

        /// <summary>Ask human player for square coordinates.</summary>
        private static Square GetSquare() {
            while (true) {
                Console.Write("  Give disk position (x,y): ");
                string input = Console.ReadLine();
                if (input != null) {
                    string[] values = input.Trim().Split(',');
                    if (values.Length == 2
                        && int.TryParse(values[0], NumberStyles.None, CultureInfo.InvariantCulture, out int x)
                        && int.TryParse(values[1], NumberStyles.None, CultureInfo.InvariantCulture, out int y)
                        && x >= 0
                        && y >= 0
                    ) {
                        return new Square(x, y);
                    }
                }
                ColorPrint.PrintError("  Give coordinates in the form 'x,y'!");
            }
        }

        /// <summary>Return player type description string.</summary>
        public string TypeString() {
            return _playerType.TypeString();
        }

        public override string ToString() {
            return $"{Disk.DiskString()} | {TypeString()} | Moves: {RoundsPlayed}";
        }
    }

    /// <summary>Player can be controlled either by a human or computer.</summary>
    public enum PlayerType {
        Human,
        Computer,
    }

    public static class PlayerTypeExtensions {
        extension(PlayerType playerType) {
            /// <summary>Check if the player is controlled by a human.</summary>
            public bool Human() {
                return playerType == PlayerType.Human;
            }

            /// <summary>Check if the player is controlled by a computer.</summary>
            public bool Computer() {
                return playerType == PlayerType.Computer;
            }

            /// <summary>Return player type description string.</summary>
            public string TypeString() {
                return playerType == PlayerType.Human ? "Human   " : "Computer";
            }
        }
    }
}
